src/iter/walk_tree.rs

```diff
@@ -0,0 +1,533 @@
+                    seen: Vec::new(),
+                    breed: self.breed,
+                    phantom: PhantomData,
+                }
+            })
+            .or_else(|| {
+                // we can still try to divide 'seen'
+                let right_seen = split_vec(&mut self.seen);
+                right_seen.map(|s| WalkTreePrefixProducer {
+                    to_explore: Default::default(),
+                    seen: s,
+                    breed: self.breed,
+                    phantom: PhantomData,
+                })
+            });
+        (self, right)
+    }
+    fn fold_with<F>(mut self, mut folder: F) -> F
+    where
+        F: Folder<Self::Item>,
+    {
+        // start by consuming everything seen
+        for s in self.seen {
+            folder = folder.consume(s);
+            if folder.full() {
+                return folder;
+            }
+        }
+        // now do all remaining explorations
+        while let Some(e) = self.to_explore.pop() {
+            self.to_explore.extend((self.breed)(&e));
+            folder = folder.consume(e);
+            if folder.full() {
+                return folder;
+            }
+        }
+        folder
+    }
+}
+
+/// ParallelIterator for arbitrary tree-shaped patterns.
+/// Returned by the [`walk_tree_prefix()`] function.
+///
+/// [`walk_tree_prefix()`]: fn.walk_tree_prefix.html
+#[derive(Debug)]
+pub struct WalkTreePrefix<S, B, I> {
+    initial_state: S,
+    breed: B,
+    phantom: PhantomData<I>,
+}
+
+impl<S, B, I> ParallelIterator for WalkTreePrefix<S, B, I>
+where
+    S: Send,
+    B: Fn(&S) -> I + Send + Sync,
+    I: IntoIterator<Item = S> + Send,
+{
+    type Item = S;
+    fn drive_unindexed<C>(self, consumer: C) -> C::Result
+    where
+        C: UnindexedConsumer<Self::Item>,
+    {
+        let producer = WalkTreePrefixProducer {
+            to_explore: once(self.initial_state).collect(),
+            seen: Vec::new(),
+            breed: &self.breed,
+            phantom: PhantomData,
+        };
+        bridge_unindexed(producer, consumer)
+    }
+}
+
+/// Create a tree-like prefix parallel iterator from an initial root node.
+/// The `children_of` function should take a node and return an iterator over its child nodes.
+/// The best parallelization is obtained when the tree is balanced
+/// but we should also be able to handle harder cases.
+///
+/// # Ordering
+///
+/// This function guarantees a prefix ordering. See also [`walk_tree_postfix`],
+/// which guarantees a postfix order.
+/// If you don't care about ordering, you should use [`walk_tree`],
+/// which will use whatever is believed to be fastest.
+/// Between siblings, children are reduced in reverse order -- that is, the children that returned last are reduced first.
+/// For example a perfect binary tree of 7 nodes will reduced in the following order:
+///
+/// ```text
+///      1
+///     / \
+///    /   \
+///   5     2
+///  / \   / \
+/// 7   6 4   3
+/// ```
```

Comment:
I was honestly confused at first how this tree was supposed to tell me the order, especially since the numbers are not organized like the values of a normally-ordered binary tree. Then I realized that they *aren't* values at all, rather the sequence indices of the traversal order, but I think this needs clarification somehow.

The other tree examples do have values, not indices, so this is inconsistent. I think it would be more clear if you first showed some tree of values, then a flat list of those values in their visited order.